docs/designs/keras_callback.md

```diff
@@ -0,0 +1,208 @@
+# Design to Support Keras Callback
+
+This document describes the design for supporting callback to customize the behavior of model during training, evaluation and inference in ElasticDL.
+
+## Motivation
+
+In deep learning, we generally need to customize the behavior of model during training, evaluation or
+inference, including reading/changing the model. According to the call frequency, we may perform the behavior per batch, per epoch or at the start and end of job. `tf.keras.callbacks.Callback` is an abstract base class and has methods to perform the behavior at different call frequency, such as `on_bath_end`, `on_epoch_end` and so on. So, we adopt the interfaces of `tf.keras.callbacks.Callback` for users to customize the behavior of model in ElasticDL. 
+
+Now we have implemented some modules which are similar to callback in ElasticDL such as `LearingRateScheduler` and `PredictionOutputsProcessor`. And user should write definitions
+in the model definition file for each module like:
+```python
+def custom_model():
+    ...
+
+# Adjust the learning rate according to iteration steps
+def learning_rate_scheduler(model_version):
+    if model_version < 5000:
+        return 0.0003
+    elif model_version < 12000:
+        return 0.0002
+    else:
+        return 0.0001
+
+# Process the prediction outputs for each batch
+class PredictionOutputsProcessor(BasePredictionOutputsProcessor):
+    ...
+```
+There will be multiple definition APIs for users to define different behaviors of model and the interfaces of APIs may be different. It is more convenient for users to define those behaviors using `tf.keras.callbacks.Callback`. 
+
+Some use cases we observed that we want to support are:
+* Case 1: Callback similar to `PredictionOutputsProcessor` that is executed after prediction outputs are made.
+* Case 2: Callback to modulate learning rate as we are currently doing in `LearningRateScheduler`.
+* Case 3: Callback to checkpoint model weights during training.
+* Case 4: Callback to write additional summaries to TensorBoard after each evaluation job completes.
+* Case 5: Callback to perform early stopping when certain conditions are met. For example, the metrics are met after an evaluation job.
+* Case 6: Callback to upload model to remote storage after completing training.
```

Comment:
For the model uploading work, we need the model export path in glusterfs and the target path in oss. How can we get these parameters from the method on_train_begin(self, logs=None)?

Reply:
There are `model` and `params` attributes in `tf.keras.callbacks.Callback`. We can default set some arguments by [set_params](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback#set_params) in `params` and user get it by `self.params`.

Reply:
For example:
```python
class UploadModelCallback(tf.keras.callbacks.Callback):
    def __init__(self, remote_url):
        self.remote_url = remote_url
        super(UploadModelCallback, self).__init__()

    def on_train_end(self, logs=None):
        saved_model_path = self.params["saved_model_path"]
        if(
            not os.path.exists(saved_model_path)
            or not os.listdir(saved_model_path)
        ):
            return
        upload(save_model_path, remote_url)
```